### PR TITLE
ALL-3936/Error listeners

### PR DESCRIPTION
**Pull request by Mecharyry**
## Problem
We created a `LocalDrmSession` but we ignored bubbling any errors up to our `no-player` error listeners. 

## Solution
Create a `DrmSessionErrorForwarder` to transform and forward our `ExoPlayer` DRM error into a `no-player` specific `Player.Error`.

We had `UUID`in three or four places for `Widevine`, we have now exposed it in a common interface. 

### Test(s) added 
No, we would need to wrap a lot of `ExoPlayer` classes in order to test the DRM part. We will do this at the end.

### Screenshots
No UI changes.

### Paired with 
@ouchadam

### CODE REVIEW

File: core/src/main/java/com/novoda/noplayer/drm/DownloadDrmSessionCreator.java [hunk 1]
```diff
@@ -1,14 +1,11 @@
...
 
+import com.google.android.exoplayer2.drm.DefaultDrmSessionManager;
 import com.google.android.exoplayer2.drm.DrmSessionManager;
 import com.google.android.exoplayer2.drm.FrameworkMediaCrypto;
 
-import java.util.UUID;
-
 class DownloadDrmSessionCreator implements DrmSessionCreator {
 
-    private static final UUID WIDEVINE_MODULAR_UUID = new UUID(0xEDEF8BA979D64ACEL, 0xA3C827DCD51D21EDL);
```
Comment (zegnus) [inline core/src/main/java/com/novoda/noplayer/drm/DownloadDrmSessionCreator.java]: 👍

File: core/src/main/java/com/novoda/noplayer/drm/DrmSessionCreator.java [hunk 2]
```diff
@@ -2,13 +2,18 @@
...
 
+import com.google.android.exoplayer2.drm.DefaultDrmSessionManager;
 import com.google.android.exoplayer2.drm.DrmSessionManager;
 import com.google.android.exoplayer2.drm.FrameworkMediaCrypto;
 
+import java.util.UUID;
+
 public interface DrmSessionCreator {
 
+    UUID WIDEVINE_MODULAR_UUID = new UUID(0xEDEF8BA979D64ACEL, 0xA3C827DCD51D21EDL);
```
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/drm/DrmSessionCreator.java]: 💯

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerDrmSessionEventListener.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package com.novoda.noplayer.exoplayer.forwarder;
+
+import com.google.android.exoplayer2.drm.DefaultDrmSessionManager;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class ExoPlayerDrmSessionEventListener implements DefaultDrmSessionManager.EventListener {
```
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerDrmSessionEventListener.java]: I think we were naming classes like this as `Listeners` ? but I am not sure if that was for the class or for the field when using it

File: core/src/main/java/com/novoda/noplayer/drm/LocalDrmSessionManager.java [hunk 4]
```diff
@@ -17,14 +19,27 @@
...
 
-    private static final UUID WIDEVINE_MODULAR_UUID = new UUID(0xEDEF8BA979D64ACEL, 0xA3C827DCD51D21EDL);
-
     private final KeySetId keySetIdToRestore;
     private final ExoMediaDrm<FrameworkMediaCrypto> mediaDrm;
+    private final DefaultDrmSessionManager.EventListener eventListener;
+    private final UUID drmScheme;
+    private final Handler handler;
+
+    static LocalDrmSessionManager newInstance(KeySetId keySetIdToRestore, ExoMediaDrm<FrameworkMediaCrypto> mediaDrm, DefaultDrmSessionManager.EventListener eventListener, UUID drmScheme) {
```
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/drm/LocalDrmSessionManager.java]: I think the `UUID` parameter would feel better next to the `ExoMediaDrm` and leaving the listener for the last. Same for the constructor. 

Of course this is not a blocker

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/DrmSessionErrorForwarder.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    private final ErrorListeners errorListeners;
+
+    DrmSessionErrorForwarder(Player player, ErrorListeners errorListeners) {
+        this.player = player;
+        this.errorListeners = errorListeners;
+    }
+
+    @Override
+    public void onDrmKeysLoaded() {
+        // TODO: Are we interested?
```
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/DrmSessionErrorForwarder.java]: @ouchadam what would be nice use case of this not-used callbacks? 

If there are none I would suggest replacing the TODO with a `not-used` comment

Comment (ouchadam) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/DrmSessionErrorForwarder.java]: currently all the `*forwarders` have a todo mentioning if we should track them, would prefer to do them all at the same time

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerDrmSessionEventListener.java (hunk 3 above)
Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerDrmSessionEventListener.java]: These are the internal listeners that are attached to either `exo-player` or `media-player`. The ones you are thinking of are the ones exposed on the public api.

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/DrmSessionErrorForwarder.java (hunk 5 above)
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/DrmSessionErrorForwarder.java]: 👍

File: core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerDrmSessionEventListener.java (hunk 3 above)
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/exoplayer/forwarder/ExoPlayerDrmSessionEventListener.java]: 👍

File: core/src/main/java/com/novoda/noplayer/drm/LocalDrmSessionManager.java (hunk 4 above)
Comment (ouchadam) [inline core/src/main/java/com/novoda/noplayer/drm/LocalDrmSessionManager.java]: 7e59b0e

**Pull request merged**